File: 南开大学/提交作业/并行程序设计/源文件/Spark/WordCount.scala
package com.example.springbootdemo.spark.wordcount

import org.apache.hadoop.fs.Path
import org.apache.spark.{SparkConf, SparkContext}

/**
 * @author youxuehu
 * 统计 word count
 */
// --input "hdfs://localhost:9000/Spark/Input/WordCount/*/"
// --output "hdfs://localhost:9000/Spark/Output/WordCount"
// spark-submit --class org.apache.spark.examples.SparkPi --master local examples/jars/spark-examples_2.11-2.0.2.jar
// spark-submit --class org.apache.spark.examples.SparkPi --master spark://localhost:7077 examples/jars/spark-examples_2.11-2.0.2.jar
// spark-submit --class org.apache.spark.examples.SparkPi --master yarn-cluster examples/jars/spark-examples_2.11-2.0.2.jar
// spark-submit --class org.apache.spark.examples.SparkPi --master yarn-client examples/jars/spark-examples_2.11-2.0.2.jar
object WordCount {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
//    conf.setMaster("local")
    conf.setAppName("wordcount")
    val sc = new SparkContext(conf)
    var input = "hdfs://localhost:9000/Spark/Input/WordCount/*"
    var output = "hdfs://localhost:9000/Spark/Output/WordCount"
    // 接收输入参数
    if (args.length == 1) {
      input = args(0)
    }
    if (args.length == 2) {
      input = args(0)
      output = args(1)
    }
    val rdd = sc.textFile(input)
    val newRdd = rdd.repartition(1)
    // 计算文章中单词出现的个数，并按照个数倒叙排序, 截取前10条数据
    val data = newRdd.flatMap(_.split(" "))
      .map((_,1)) // map  输出
      .reduceByKey(_+_) // reduce计算
      .sortBy(_._2,false) // sort排序
      .repartition(1)
      .map{x =>
            x._1 + "\t" + x._2
      }
    val hadoopConf = sc.hadoopConfiguration
    val hdfs = org.apache.hadoop.fs.FileSystem.get(hadoopConf)
    val path = new Path(output);
    if(hdfs.exists(path)){
      // 如果输出目录存在，则删除
      hdfs.delete(path)
    }
    // data.foreach(println)
    // 将list转换成rdd
    val new_data = sc.parallelize(data.collect())
    // 最终结果写入HDFS
    new_data.saveAsTextFile(output)
  }
}

File: 南开大学/提交作业/并行程序设计/源文件/Spark/WordCount.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
from pyspark import SparkConf, SparkContext
import sys
if __name__ == "__main__":
    if len(sys.argv) != 3:
        print "Usage: wordcount <file>"
        sys.exit(-1)
    sparkConf = SparkConf()\
        .setAppName('Python Spark WordCount')\
        .setMaster('local[2]')
    sc = SparkContext(conf=sparkConf)
    sc.setLogLevel('WARN')
    rdd = sc.textFile(sys.argv[1])
    word_count_rdd = rdd\
            .flatMap(lambda line: line.split(" "))\
                     .map(lambda word: (word, 1))\
                     .reduceByKey(lambda a, b: a + b)
    word_count_rdd.repartition(1).saveAsTextFile(sys.argv[2])
    sc.stop()
